Refactored router and dispatcher middleware to be more like Express.

imalberto opened the pull request:
- updated version to 0.8.x  to keep separate from #develop branch
- `routes.json` configuration is no longer contextualized - always default to "master"
- refactored `mojito-router` middleware 
- refactored `mojito-handler-dispatcher` middleware
- updated `ResourceStore.getRoutes()` to be not context aware
- refactored module `route-maker` to expand the route config
# todo
- [x] fix unit tests
- [x] fix functional tests
- [x] fix examples

imalberto committed:
First pass at refactoring dispatcher

imalberto committed:
Updated dispatcher and router.

imalberto committed:
Deleted dispather and router middleware as they have been moved into their own submodule

imalberto committed:
Removed unused code

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -131,6 +142,7 @@ YUI.add('mojito-route-maker', function(Y, NAME) {
             }
 
             if (typeof route.params !== 'object') {
+                Y.log('** Deprecated ** params: "name=Tom&age=23" is no longer supported', 'info');
```
should be `warn` instead of `info`, and make sure you also add `NAME` as the 3rd argument for better filtering.

caridy commented on lib/app/autoload/store.server.js:
```diff
@@ -772,14 +772,17 @@ YUI.add('mojito-resource-store', function(Y, NAME) {
             var appConfig,
                 cachedValue,
-                key = JSON.stringify(ctx || {}),
+                ctx = {}, // static context
```
not needed anymore

caridy commented on lib/app/autoload/store.server.js:
```diff
@@ -772,14 +772,17 @@ YUI.add('mojito-resource-store', function(Y, NAME) {
-        getRoutes: function(ctx) {
+        getRoutes: function () {
             var appConfig,
                 cachedValue,
```
is this needed?

caridy commented on lib/dispatcher.js:
```diff
@@ -0,0 +1,79 @@
+    @param {http.ServerResponse} res
+    @param {Function} next
+    **/
+    dispatch: function (req, res, next) {
```
This is not a middleware, it should not have a middleware signature. lets discuss the details offline.

caridy commented on lib/dispatcher.js:
```diff
@@ -0,0 +1,79 @@
+        outputHandler.page.appConfig = store.getAppConfig(context);
+
+        // compute routes once per request
+        outputHandler.page.routes = store.getRoutes();
```
why do we need the routes attached to page anymore?

caridy commented on lib/mojito.js:
```diff
@@ -113,6 +115,7 @@ Mojito.prototype._init = function (app) {
     options.context = { runtime: 'server' }; // TODO: use correct static context here
     options.mojitoRoot = __dirname;
     options.root = root;
+    console.log(process.cwd());
```
can we use `debug()` instead?

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,177 @@
+
+    app = express();
+    ...
+    app.use(mojito.attachRoutes());
```
`attachRoutes` should NOT be a middleware because it will force to register the routes once the first request comes in.

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,177 @@
+                    if (route.verbs.hasOwnProperty(verb)) {
+                        // debug('instaling handler: [' + verb + '] ' + name);
+                        verb = verb.toLowerCase();
+                        app[verb](new RegExp(route.ext_match), handler(name, route));
```
why are all routes regex?

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,177 @@
+                route = routes[name];
+                for (verb in route.verbs) {
+                    if (route.verbs.hasOwnProperty(verb)) {
+                        // debug('instaling handler: [' + verb + '] ' + name);
```
why are all `debug` statements commented out?

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,177 @@
+         * @param {Object} route route configuration
+         * @return {Function} express.middleware
+         */
+        function handler(name, route) {
```
This should be exposed, so people can do this in their app.js:

```
app.get('/foo', mojito.dispatch('foo.index'));
```

which is going to be equivalent to do this in routes.json:

```
{
    "foo.index": {
         "path": "/foo",
         "verb": ["get"],
         "call": "foo.index"
    }
}
```

which also means, you will be able to do this in your templates: `{{#linkTo "foo.index"}}`, although this is not part of this PR, but the next one.

caridy commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
     "name": "mojito",
-    "version": "0.6.1",
+    "version": "0.10.1",
```
why version 10?

caridy commented:
finished the first pass, let's talk offline for more details.

imalberto committed:
Updated with PR feedback

imalberto committed:
Linted

imalberto committed:
Added backward compatibility feature for route.params.

imalberto committed:
Marking `doCallReplacement` and `expand` for deletion.

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,220 @@
+                            args.push(mid(route.params));
+                        }
+
+                        args.push(dispatcher.dispatch(route));
```
`route.call`

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,220 @@
+
+                        args = [route.path];
+
+                        if (route.params) {
```
nice! remember to add a BIG warn for deprecation.

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,220 @@
+                    query;
+
+                if (!copy) {
+                    // easier way ?
```
`var util = require('./path/to/lib/util');`

that file should hold any utility method needed for the server side (hence anything needed in a node module), feel free to add `copy` method in there.

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,220 @@
+                query = req.routeMatch.query;
+
+                cloned = copy(o);
+                debug(inspect(cloned));
```
remove!

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,220 @@
+                for (key in cloned) {
+                    if (cloned.hasOwnProperty(key) && !query[key]) {
+                        debug('copying params from routes configuration: %s', key);
+                        query[key] = cloned[key];
```
I thought those params defined per route in routes.json were intended to be just attached as req.params under the hood. In which case all these cloning and stuff are not very useful.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -156,33 +170,33 @@ YUI.add('mojito-route-maker', function(Y, NAME) {
-        });
-        return route;
-    };
+    // doCallReplacement = function (route, uri) {
```
if we don't need it, remove it :)

caridy commented:
A lot of comments, let's do another pass tomorrow

imalberto committed:
- deleted unused code blocks
- updated based on PR review
- added `readConfig` to lib/util.js to read configuration (either JSON or YAML format) . probably the rs.config add-on can be refactored to use this.

imalberto committed:
Restored `page.routes` as its required on `ac.url` add-on.

imalberto committed:
- Restored `route-maker` as `ac.url` add-on depends on it.

imalberto commented on lib/dispatcher.js:
```diff
@@ -0,0 +1,132 @@
+        outputHandler.page.staticAppConfig = appConfig;
+        outputHandler.page.appConfig = store.getAppConfig(context);
+        // routes are not contextualized anymore
+        outputHandler.page.routes = store.getRoutes();
```
For now, this is required by `ac.config` and `ac.url`

imalberto committed:
Fixed typos in log messages.

caridy commented on lib/dispatcher.js:
```diff
@@ -0,0 +1,132 @@
+
+        return function (req, res, next) {
+            var command = { instance: { } },
+                mojito = (req.app && req.app.mojito) || null,
```
not need to use `|| null` here.

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,237 @@
+
+                // TODO: find a more appropriate namespace for this
+                req.routeMatch = req.routeMatch || {};
+                req.routeMatch.query = req.routeMatch.query || {};
```
how is this used across mojito? who uses those query params added manually from routes.json?

caridy commented:
we are getting close :)

imalberto committed:
Fixed the namespace object on `req` to stash route specific configuration.

imalberto committed:
- Fixed log message
- Removed `route.params` support

imalberto committed:
`route.params` is now set from `req.params`

imalberto committed:
Removed unused code block.

imalberto committed:
Updated docs

caridy commented on lib/dispatcher.js:
```diff
@@ -0,0 +1,132 @@
+        var my = this,
+            matchCall = [];
+
+        matchCall[0] = call.split('.');
```
this is a little confusing code. Overloading `matchCall[0]` is confusing. Either you add a good note, or change the code with more variables to avoid the overloading.

caridy commented on lib/dispatcher.js:
```diff
@@ -0,0 +1,132 @@
+    **/
+    handleRequest: function (req, res, next) {
+
+        var command = req.command || null,
```
`|| null` is really not needed.

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,216 @@
+                    route = routes[name];
+                    for (verb in route.verbs) {
+                        if (route.verbs.hasOwnProperty(verb)) {
+                            debug('[%s %s] installing handler', name, route.path);
```
I don't think the name will be relevant! the `call` and `path` are!

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,216 @@
+         *
+         * TODO: should this be a separate call that `app.js` should do ?
+         */
+        function registerTunnelRpc() {
```
I don't think this belongs to here! Tunnel should be either: 

a) set in app.js as a middleware with path: e.g.: `app.post('/tunnel', tunnelMiddleware);` or
b) mojito.js can do it automatically which is not good because they devs will not have a way to change it or modify it. 

but it is definitely not related to router.js

caridy commented on lib/router.js:
```diff
@@ -0,0 +1,216 @@
+            }
+        }
+
+        registerTunnelRpc();
```
same as above!

caridy commented:
done with the pass!

imalberto committed:
Updated with feedback from PR:
- move roc tunnel registration to its own submodule
- updated docs
- simplify code blocks for clarity

caridy commented:
+1

imalberto committed:
Added tunnel submodule.

imalberto committed:
Updated api docs.

imalberto committed:
- added unit tests for lib/router and lib/dispatcher
- linted <3

imalberto committed:
Fixed the following functional tests:
- tests/common/*
- tests/func/examples/*
- disabled some tests from tests/func/common/common_descriptor.json and moved them to disabled-common_descriptor.txt for revisiting later.

imalberto committed:
Fixed the following func tests:
- tests/func/examples/*
- tests/func/usecases

imalberto committed:
Fix fund tests :
- tests/func/serveronly

imalberto committed:
Fixed fund tests:
- tests/func/routing.

Disabled the following tests for now:
- testbasicroutingneg3
- testbasicrouting10-1
- testbasicrouting10-2
- testbasicrouting11neg1
- testbasicrouting11neg1
- testbasicrouting11pos

imalberto committed:
Fix func test quickstartguide.

imalberto committed:
Fix func tests tests/func/common/*

imalberto committed:
Restored some of the functional tests under tests/func/common/*

imalberto committed:
Linted + fixes to app.js

imalberto committed:
Linted

imalberto committed:
Removed disabled-common_descriptor.txt, since all of the functional tests are now working

caridy commented:
+1

drewfish commented on lib/dispatcher.js:
```diff
@@ -0,0 +1,136 @@
+        // Y.mojito.hooks.hook('AppDispatch', outputHandler.hook, req, res);
+        // HookSystem::EndBlock
+
+        Y.mojito.Dispatcher.init(store).dispatch(command, outputHandler);
```
Do we need to initialized the dispatcher for every request?

caridy replied:
hopefully not! lets double check that to see how expensive is this.

aljimenez commented:
Is there a reason why routes are no longer contextualized? Currently all routes in routes.json are getting exposed to the client.

caridy commented:
Yes, the reason is perf. For the routes to be contextualize, we need to process them per request, this is an expensive operation, and will now allow us to rely on the internal cache implemented by express. Also, the fact that routes are normally not changed per context.

On top of that, this doesn't necessarily means you cannot limit what you expose to the client side. In fact, you have full control of what is exposed, by using express-map's filter mechanism to expose the list of routes you want, in case you don't want them all to be exposed.
